libs/langchain/langchain/vectorstores/redis/base.py

```diff
@@ -920,6 +922,64 @@ def similarity_search(
             )
         return docs
 
+    def max_marginal_relevance_search(
+        self,
+        query: str,
+        k: int = 4,
+        fetch_k: int = 20,
+        lambda_mult: float = 0.5,
+        filter: Optional[RedisFilterExpression] = None,
+        return_metadata: bool = True,
+        distance_threshold: Optional[float] = None,
+        **kwargs: Any,
+    ) -> List[Document]:
+        """Return docs selected using the maximal marginal relevance.
+
+        Maximal marginal relevance optimizes for similarity to query AND diversity
+            among selected documents.
+
+        Args:
+            query (str): Text to look up documents similar to.
+            k (int): Number of Documents to return. Defaults to 4.
+            fetch_k (int): Number of Documents to fetch to pass to MMR algorithm.
+            lambda_mult (float): Number between 0 and 1 that determines the degree
+                of diversity among the results with 0 corresponding
+                to maximum diversity and 1 to minimum diversity.
+                Defaults to 0.5.
+            filter (RedisFilterExpression, optional): Optional metadata filter.
+                Defaults to None.
+            return_metadata (bool, optional): Whether to return metadata.
+                Defaults to True.
+            distance_threshold (Optional[float], optional): Distance threshold
+                for vector distance from query vector. Defaults to None.
+
+        Returns:
+            List[Document]: A list of Documents selected by maximal marginal relevance.
+        """
+        # Fetch the initial documents
+        prefetch_docs = self.similarity_search(
+            query=query,
+            k=fetch_k,
+            filter=filter,
+            return_metadata=return_metadata,
+            distance_threshold=distance_threshold,
+            **kwargs,
+        )
+
+        # Embed the query and the fetched documents
+        prefetch_embeddings = [
+            self._embeddings.embed_query(doc.page_content) for doc in prefetch_docs
+        ]
+        prefetch_embeddings = np.array(prefetch_embeddings)
+        query_embedding = self._embeddings.embed_query(query)
```

Comment:
This is embedding the query twice, once inside `similarity_search` and a 2nd time in this line, we should do it only once